data_structures/src/fee.rs

```diff
@@ -0,0 +1,205 @@
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl Zero for AbsoluteFee {
+    #[inline]
+    fn zero() -> Self {
+        Self(Wit::zero())
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Hash, PartialEq, Eq, PartialOrd, Serialize)]
+pub struct RelativeFee(Priority);
+
+impl RelativeFee {
+    #[inline]
+    pub fn into_absolute(self, weight: u32) -> AbsoluteFee {
+        AbsoluteFee(self.0.derive_fee_wit(weight))
+    }
+}
+
+impl fmt::Display for RelativeFee {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} nWitWu", self.0.as_f64())
+    }
+}
+
+impl str::FromStr for RelativeFee {
+    type Err = <f64 as str::FromStr>::Err;
+
+    #[inline]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        f64::from_str(s).map(Priority::from).map(Self)
+    }
+}
+
+impl ops::Add for RelativeFee {
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl num_traits::Zero for RelativeFee {
+    #[inline]
+    fn zero() -> Self {
+        Self(Priority::zero())
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Hash, PartialEq, Eq, PartialOrd, Serialize)]
+/// Type for representing a fee value that can be absolute (nanoWits) or relative (priority).
+pub enum Fee {
+    /// An absolute fee, as expressed in nanoWits.
+    Absolute(AbsoluteFee),
+    /// A relative fee, aka "priority", as expressed as nanoWits (or fractional amounts) per weightunit.
+    Relative(RelativeFee),
+}
+
+impl Fee {
+    #[inline]
+    pub fn absolute_from_wit(wit: Wit) -> Self {
+        Self::Absolute(AbsoluteFee(wit))
+    }
+
+    #[inline]
+    pub fn absolute_from_nanowits(nanowits: u64) -> Self {
+        Self::absolute_from_wit(Wit::from_nanowits(nanowits))
+    }
+
+    pub fn relative_from_float<T>(float: T) -> Self
+    where
+        f64: From<T>,
+    {
+        Self::Relative(RelativeFee(Priority::from(f64::from(float))))
+    }
+}
+
+impl Default for Fee {
+    fn default() -> Self {
+        <Self as Zero>::zero()
+    }
+}
+
+impl fmt::Display for Fee {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Fee::Absolute(absolute) => absolute.fmt(f),
+            Fee::Relative(relative) => relative.fmt(f),
+        }
+    }
+}
+
+impl ops::Add for Fee {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use crate::fee::Fee::*;
+        match (self, rhs) {
+            (Absolute(lhs), Absolute(rhs)) => Fee::Absolute(lhs + rhs),
+            (Relative(lhs), Relative(rhs)) => Fee::Relative(lhs + rhs),
+            _ => {
+                unimplemented!()
+            }
+        }
+    }
+}
```

Comment:
If this impl impl didn't exist, adding an absolute `Fee` with a relative `Fee` would be a compile time error. So I would remove it.

Reply:
I added this `impl` because `ops::Add` is a requirement for `num_traits::Zero`. Shall I get rid of that and relay on the `Default` trait instead? I sort of like `Zero` more because of its unambiguous semantics.

Reply:
Yes, remove the `Zero` as well.

Reply:
Out it goes!